Bug 2010365: OpenShift Alerting Rules Style-Guide Compliance

Davoska opened the pull request:
Add missing namespace labels `openshift-cluster-version` to alerting rules to comply with 
the [style guidance](https://github.com/openshift/enhancements/blob/master/enhancements/monitoring/alerting-consistency.md#style-guide). Alerts should include a namespace label indicating the alert's source.

Davoska commented:
/retest-required

wking commented on install/0000_90_cluster-version-operator_02_servicemonitor.yaml:
```diff
@@ -56,6 +57,7 @@ spec:
       expr: |
         (time()-cluster_version_operator_update_retrieval_timestamp_seconds) >= 3600 and ignoring(condition, name, reason) cluster_operator_conditions{name="version", condition="RetrievedUpdates", endpoint="metrics", reason!="NoChannel"}
       labels:
+        namespace: openshift-cluster-version
```
Checking Telemetry, `CannotRetrieveUpdates` seems to include `namespace` already via `cluster_version_operator_update_retrieval_timestamp_seconds` and `cluster_operator_conditions` scraping.  So I'd rather drop this line from the commit as redundant.

Davoska replied:
Thank you, fixed in https://github.com/openshift/cluster-version-operator/pull/800/commits/d525eb97a2c971b4837210963d86d3b617452891.

wking commented on install/0000_90_cluster-version-operator_02_servicemonitor.yaml:
```diff
@@ -64,6 +66,7 @@ spec:
       expr: |
         sum by (channel,upstream) (cluster_version_available_updates) > 0
       labels:
+        namespace: openshift-cluster-version
```
`UpdateAvailable` already has `namespace` in Telemetry (this time via `cluster_version_available_updates`), although we probably lost that with 0721e7bfa (#643).  Can we add `namespace` to the `sum by` instead, with:

```
sum by (channel,namespace,upstream) (cluster_version_available_updates) > 0
```

and drop this line?

Davoska replied:
Thank you, fixed in https://github.com/openshift/cluster-version-operator/pull/800/commits/d525eb97a2c971b4837210963d86d3b617452891.

wking commented on install/0000_90_cluster-version-operator_02_servicemonitor.yaml:
```diff
@@ -75,6 +78,7 @@ spec:
         max by (name, condition, endpoint) (cluster_operator_conditions{name="version", condition="Upgradeable", endpoint="metrics"} == 0)
       for: 60m
       labels:
+        namespace: openshift-cluster-version
```
Here's another case where I'd rather include `namespace` in the aggregation (`max by (namespace, name, condition, endpoint) (...)`), instead of clearing the label and then injecting a hard-coded replacement.

Davoska replied:
Thank you, fixed in https://github.com/openshift/cluster-version-operator/pull/800/commits/d525eb97a2c971b4837210963d86d3b617452891.

wking commented on install/0000_90_cluster-version-operator_02_servicemonitor.yaml:
```diff
@@ -84,6 +88,7 @@ spec:
         cluster_operator_up{job="cluster-version-operator"} == 0
       for: 10m
       labels:
+        namespace: openshift-cluster-version
```
`ClusterOperatorDown` already has `namespace` in Telemetry (this time via `cluster_operator_up`), so I'd rather drop this line from the commit as redundant.

We could consider a `max by (namespace, name, ...) (...)` aggregation, if we wanted to exclude `pod` to avoid alert churn whenever the CVO pod cycled.  But we're not all that consistent on this front today, so if we do decide to do this sort of thing, we should probably put in enough thought to make the change consistently (e.g. I'm not sure why we are including `endpoint` but not `service` in `ClusterNotUpgradeable` today).

Davoska replied:
Thank you, fixed in https://github.com/openshift/cluster-version-operator/pull/800/commits/d525eb97a2c971b4837210963d86d3b617452891.

I have also added the aggregation operator `max by` to `CannotRetrieveUpdates`, `ClusterOperatorDown`, `ClusterOperatorDegraded`, and `ClusterOperatorFlapping` to avoid the alert churn whenever the CVO pod cycles in the new commit https://github.com/openshift/cluster-version-operator/pull/800/commits/2b8ef2a4c653df59cf67991b5ecc8c827767fe91.

I am not sure of the purpose of the `endpoint` or potentially `service` in `ClusterNotUpgradeable`, so I have not made this modification.

wking commented on install/0000_90_cluster-version-operator_02_servicemonitor.yaml:
```diff
@@ -97,6 +102,7 @@ spec:
         ) == 1
       for: 30m
       labels:
+        namespace: openshift-cluster-version
```
`ClusterOperatorDegraded` already has `namespace` in Telemetry (this time via `cluster_operator_conditions`, so I'd rather drop this line from the commit as redundant.  We might need to adjust the `cluster_operator_up` to preserve `namespace` in that case.  Something like:

```
cluster_operator_conditions{job="cluster-version-operator", condition="Degraded"}
or on (namespace, name)
group by (namespace, name) (cluster_operator_up{job="cluster-version-operator"})
```

Davoska replied:
Thank you, fixed in https://github.com/openshift/cluster-version-operator/pull/800/commits/d525eb97a2c971b4837210963d86d3b617452891.

wking commented on install/0000_90_cluster-version-operator_02_servicemonitor.yaml:
```diff
@@ -106,4 +112,5 @@ spec:
         changes(cluster_operator_up{job="cluster-version-operator"}[2m]) > 2
       for: 10m
       labels:
+        namespace: openshift-cluster-version
```
`ClusterOperatorFlapping` already has `namespace` in Telemetry (this time via `cluster_operator_up`), so I'd rather drop this line from the commit as redundant.

Davoska replied:
Thank you, fixed in https://github.com/openshift/cluster-version-operator/pull/800/commits/d525eb97a2c971b4837210963d86d3b617452891.

Davoska commented:
/hold addressing comments

Davoska commented:
Thank you, Trevor, for the explanations. It seems that I forgot about the purpose of aggregation operators and receiving labels implicitly from metrics :facepalm:. I wanted to be explicit about the alerts having the namespace. However, I now understand that it's redundant. 

Fixed the redundant labels `namespace` in the new commit https://github.com/openshift/cluster-version-operator/pull/800/commits/d525eb97a2c971b4837210963d86d3b617452891.

Added aggregation operators to some alerts to exclude necessary labels (mainly for the purpose of avoiding multiple alerts when the CVO pod cycles) in the new commit https://github.com/openshift/cluster-version-operator/pull/800/commits/2b8ef2a4c653df59cf67991b5ecc8c827767fe91.

/unhold

Davoska commented:
https://github.com/openshift/cluster-version-operator/commit/2b8ef2a4c653df59cf67991b5ecc8c827767fe91 -> https://github.com/openshift/cluster-version-operator/commit/ffb2602fb241234480a9ecb37a23279625c97f36 wording of the commit message.

/unhold

Davoska commented:
/retest

Davoska commented:
/retest

LalatenduMohanty reviewed (approved):
/lgtm

LalatenduMohanty commented:
/hold For Trevor to take a look

wking reviewed (approved):
/lgtm
/hold cancel

Davoska commented:
/retest

wking commented:
[This run][1] passed, except for timing out during teardown.

/override ci/prow/e2e-agnostic-upgrade

[1]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-version-operator/800/pull-ci-openshift-cluster-version-operator-master-e2e-agnostic-upgrade/1555797022764699648

Davoska commented:
@wking, I guess we need to run the `/override` again? :thinking: I am not sure why it's still not passing because there weren't any merges to the `master` branch since that.

Davoska commented:
https://github.com/openshift/cluster-version-operator/pull/800#issuecomment-1208445514:
> [This run](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-version-operator/800/pull-ci-openshift-cluster-version-operator-master-e2e-agnostic-upgrade/1555797022764699648) passed, except for timing out during teardown.
> 
> /override ci/prow/e2e-agnostic-upgrade

Unholding the PR and retesting, so we can potentially override the ci/prow/e2e-agnostic-upgrade, because the head of the master branch has changed.

Davoska commented:
/unhold

Davoska commented:
/retest

wking commented:
Still hitting [orthogonal disruption failures][1] because the origin suite doesn't scale disruption for our A->B->A rollback presubmit.

/override ci/prow/e2e-agnostic-upgrade

[1]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-version-operator/800/pull-ci-openshift-cluster-version-operator-master-e2e-agnostic-upgrade/1559518659724972032
